cloudapi/logs.go

```diff
@@ -164,8 +198,17 @@ func (c *Config) StreamLogsToLogger(
 		}
 
 		if err != nil {
-			logger.WithError(err).Warn("error reading a message from the cloud")
+			logger.WithError(err).Warn("error reading a log message from the cloud, trying to establish a fresh connection with the logs service...") //nolint:lll
 
+			newconn, errd := c.logtailConn(ctx, referenceID, latest.TimeOrNow())
```

Comment:
I would argue that now the whole readMessage, send on channel, set latest timestamp based on it is racy (logically if not programmatically given the inclusion of atomics now).

I doubt it will make a difference where we get messages fast enough that we can't log them until we get the next one and then a third one is an error, so maybe not do anything, but add a comment saying that this *non* exact correctness is on purpose?

Reply:
I added the comment but I would like to have a 100% correct solution but I didn't yet find one that can maintain the code simple as-is. If you have suggestions they are welcome.